src/app/Provider.tsx

```diff
@@ -0,0 +1,22 @@
+"use client";
+
+import { ReactNode } from "react";
+import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
+import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
+
+const queryClient = new QueryClient();
```

Comment:
`Next.js`에서는 queryClient를 React.useState 내에 선언할 것을 권장하고 있습니다. 그 이유에 대해 알아보고 이해한 것은 다음과 같습니다!

- `Next.js`는 페이지 이동 시에 `layout.tsx`에서부터 새롭게 렌더링한다고 합니다.
- 이 때 컴포넌트 외부에 단순히 인스턴스를 생성하면 페이지마다 새로운 인스턴스가 생성됩니다.
- 따라서 참조 동일성을 보장할 수 없기 때문에 렌더링 성능에 영향을 미친다고 합니다.

[공식 문서 예제](https://tanstack.com/query/latest/docs/framework/react/examples/nextjs-suspense-streaming)
[참고1](https://velog.io/@cjjss11/React-Query-v5#1-reactqueryproviders-%EB%A7%8C%EB%93%A4%EA%B8%B0)

Reply:
저도 성민님 의견보고 처음 알게되서 공식문서를 찾아보니 아래와 같이 알려주네요.
- 데이터가 서로 다른 사용자와 요청 사이에서 공유되지 않도록 보장
- queryClient가 컴포넌트 생명주기에서 단 한 개만 생성
[블로그 참고 - pages router](https://velog.io/@okko8522/Next.JS%EC%97%90-%EB%A6%AC%EC%95%A1%ED%8A%B8-%EC%BF%BC%EB%A6%AC-%EC%A0%81%EC%9A%A9%ED%95%98%EA%B8%B0#%EC%B0%B8%EA%B3%A0-%EC%99%9C-queryclient-%EC%84%A4%EC%A0%95%EC%97%90-usestate%EB%A5%BC)
[Tanstack Query - pages router](https://tanstack.com/query/v5/docs/framework/react/guides/ssr#initial-setup)

근데 위 내용은 pages router를 기준이었습니다!

app router는 찾아보니 위와 다르게 useState를 사용하지 않네요..?

공식 문서에 따르면, 서버와 클라이언트를 기준으로 queryClient 생성 방식이 다르다고 합니다!
- **서버**: 항상 새로운 queryClient 생성
- **클라이언트**: 생성한 queryClient가 없다면 생성하고, 있다면 기존 queryClient 반환

```ts
// In Next.js, this file would be called: app/providers.jsx
'use client'

// We can not useState or useRef in a server component, which is why we are
// extracting this part out into it's own file with 'use client' on top
import { useState } from 'react'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

function makeQueryClient() {
  return new QueryClient({
    defaultOptions: {
      queries: {
        // With SSR, we usually want to set some default staleTime
        // above 0 to avoid refetching immediately on the client
        staleTime: 60 * 1000,
      },
    },
  })
}

let browserQueryClient: QueryClient | undefined = undefined

function getQueryClient() {
  if (typeof window === 'undefined') {
    // Server: always make a new query client
    return makeQueryClient()
  } else {
    // Browser: make a new query client if we don't already have one
    // This is very important so we don't re-make a new client if React
    // suspends during the initial render. This may not be needed if we
    // have a suspense boundary BELOW the creation of the query client
    if (!browserQueryClient) browserQueryClient = makeQueryClient()
    return browserQueryClient
  }
}

export default function Providers({ children }) {
  // NOTE: Avoid useState when initializing the query client if you don't
  //       have a suspense boundary between this and the code that may
  //       suspend because React will throw away the client on the initial
  //       render if it suspends and there is no boundary
  const queryClient = getQueryClient()

  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  )
}
```
위 내용은 공식문서 Advanced Server Rendering 파트에서 가져왔는데 고급 설정은 위와 같이 해야하고, 일반적인 설정은 useState를 따라가도 되는건지 잘 모르겠네요..

[Tanstack Query - app router](https://tanstack.com/query/latest/docs/framework/react/guides/advanced-ssr)

Reply:
여러 블로그들을 찾아보니 queryClient를 선언하고 관리하는 방법이 다양한데 이 부분도 이따 회의에서 다시 한번 이야기 해 보시죠! 추가로 성민님께서 여쭤보신 내용은 제가 해석할 때는 Suspense 바운더리 내에서 쿼리 클라이언트를 사용할 경우 초기 렌더링 시 쿼리 클라이언트가 Suspense로 인해 폐기될 수 있기에 useState로 쿼리 클라이언트를 초기화 하지 말라는 내용인 것 같은데, 이 부분도 이따 회의에서 다시 한번 얘기해보시죠 ㅎㅎ